File: packs/dsa-trees-py/challenges/02a-max-depth-recursive.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import max_depth


def test_balanced_tree():
    root = build_tree([3, 9, 20, None, None, 15, 7])
    assert max_depth(root) == 3


def test_right_skewed():
    root = build_tree([1, None, 2])
    assert max_depth(root) == 2


def test_empty_tree():
    assert max_depth(None) == 0


def test_single_node():
    root = build_tree([1])
    assert max_depth(root) == 1


def test_left_skewed():
    root = build_tree([1, 2, None, 3])
    assert max_depth(root) == 3

File: packs/dsa-trees-ts/challenges/02a-max-depth-recursive.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '02a-max-depth-recursive.json');

describe('Max Depth (Recursive)', () => {
  let maxDepth: (root: TreeNode | null) => number;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ maxDepth: typeof maxDepth }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    maxDepth = mod.maxDepth;
  });

  afterAll(() => cleanup());

  it('returns 3 for a balanced tree', () => {
    expect(maxDepth(buildTree([3, 9, 20, null, null, 15, 7]))).toBe(3);
  });

  it('returns 2 for right-skewed tree', () => {
    expect(maxDepth(buildTree([1, null, 2]))).toBe(2);
  });

  it('returns 0 for empty tree', () => {
    expect(maxDepth(null)).toBe(0);
  });

  it('returns 1 for single node', () => {
    expect(maxDepth(buildTree([1]))).toBe(1);
  });

  it('returns 3 for left-skewed tree', () => {
    expect(maxDepth(buildTree([1, 2, null, 3]))).toBe(3);
  });
});
